Mathlib/Algebra/Group/Submonoid/Pointwise.lean

```diff
@@ -588,6 +615,33 @@ theorem pow_subset_pow {s : AddSubmonoid R} {n : ℕ} : (↑s : Set R) ^ n ⊆ 
 
 end Semiring
 
+section SMul
+
+variable [AddMonoid R] [DistribSMul R A]
+
+instance : SMul (AddSubmonoid R) (AddSubmonoid A) where
```

Comment:
Is `A := AddSubmonoid R` possible here? If so, then this instance probably creates diamonds

Reply:
For `Semiring R` it's possible, but there's no diamond:
```
example {R} [Semiring R] : Mul.toSMul (AddSubmonoid R) = AddSubmonoid.smul := rfl
```
For NonUnital/NonAssocSemiring there's strangely no DistribSMul instance. For Semiring it goes through [DistribMulAction.toDistribSMul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupWithZero/Action/Defs.html#DistribMulAction.toDistribSMul) but for NonUnital/NonAssoc there's no MulAction.

I should definitely change this def to `protected def` and `scoped[Pointwise] attribute [instance]` like the others though ...

Reply:
Oh sorry I misread, you're talking about `SMul (AddSubmonoid R) (AddSubmonoid (AddSubmonoid R))` ...

Reply:
There's no existing `SMul (AddSubmonoid R) (AddSubmonoid (AddSubmonoid R))` instance, even for [CommRing R].
```
import Mathlib
open Pointwise
variable {R} [CommRing R]
#synth SMul (AddSubmonoid R) (AddSubmonoid <| AddSubmonoid R)
-- failed to synthesize AddZeroClass (AddSubmonoid R)
```
The reason is that there aren't many instances declared on [AddSubmonoid](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Submonoid/Defs.html#AddSubmonoid). There's much more instances on Submodule, e.g. [Submodule.idemSemiring](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Operations.html#Submodule.idemSemiring), so we should probably worry more about that ...